packages/fullscreen/src/index.ts

```diff
@@ -0,0 +1,154 @@
+import { useState } from "react";
+
+type EventCallback = (this: Document, ev: any) => any;
+interface NormalizedFullscreenApi {
+  requestFullscreen: string;
+  exitFullscreen: string;
+  fullscreenElement: string;
+  fullscreenEnabled: string;
+  fullscreenchange: string;
+  fullscreenerror: string;
+}
+
+const getBrowserFunctions = (): NormalizedFullscreenApi => {
+  let val;
+
+  const fnMap = [
+    [
+      "requestFullscreen",
+      "exitFullscreen",
+      "fullscreenElement",
+      "fullscreenEnabled",
+      "fullscreenchange",
+      "fullscreenerror"
+    ],
+    // New WebKit
+    [
+      "webkitRequestFullscreen",
+      "webkitExitFullscreen",
+      "webkitFullscreenElement",
+      "webkitFullscreenEnabled",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    // Old WebKit
+    [
+      "webkitRequestFullScreen",
+      "webkitCancelFullScreen",
+      "webkitCurrentFullScreenElement",
+      "webkitCancelFullScreen",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    [
+      "mozRequestFullScreen",
+      "mozCancelFullScreen",
+      "mozFullScreenElement",
+      "mozFullScreenEnabled",
+      "mozfullscreenchange",
+      "mozfullscreenerror"
+    ],
+    [
+      "msRequestFullscreen",
+      "msExitFullscreen",
+      "msFullscreenElement",
+      "msFullscreenEnabled",
+      "MSFullscreenChange",
+      "MSFullscreenError"
+    ]
+  ];
+
+  let i = 0;
+  const l = fnMap.length;
+  const ret = {} as NormalizedFullscreenApi;
+
+  for (; i < l; i++) {
+    val = fnMap[i];
+    if (val && val[1] in document) {
+      for (i = 0; i < val.length; i++) {
+        ret[fnMap[0][i]] = val[i];
+      }
+      return ret;
+    }
+  }
+};
+
+const useFullscreen = () => {
+  const [isFullscreen, setIsFullscreen] = useState(false);
+  const [element, setElement] = useState(null);
+  const fn = getBrowserFunctions();
+
+  const eventNameMap = {
+    change: fn.fullscreenchange,
+    error: fn.fullscreenerror
+  };
+
+  const request = (element?: HTMLElement) =>
+    new Promise((resolve, reject) => {
+      const onFullScreenEntered = () => {
+        off("change", onFullScreenEntered);
+        resolve();
+      };
+
+      on("change", onFullScreenEntered);
+
+      element = element || document.documentElement;
+      setIsFullscreen(true);
+      setElement(element);
+
+      Promise.resolve(element[fn.requestFullscreen]()).catch(reject);
```

Comment:
Are these methods going to work for all browsers? Seems like they're all defined in the first part of the code but not being used anywhere? Or am I missing something?

Reply:
In the `for` loop it builds the proper function map for the current browser. For example, if it was running in IE

the map might look something like this
```javascript
{
      requestFullscreen: "msRequestFullscreen",
      exitFullscreen : "msExitFullscreen",
      fullscreenElement: "msFullscreenElement",
      fullscreenEnabled : "msFullscreenEnabled",
      fullscreenchange: "MSFullscreenChange",
      fullscreenerror: "MSFullscreenError"
}
```

this way I am standardizing the API, and I can use the same function names that will map to the proper function without a bunch of conditional checks throughout the code.

Reply:
Right didn't notice that, makes perfect sense.